lib/elixir/lib/uri.ex

```diff
@@ -977,6 +977,60 @@ defmodule URI do
       %{uri | query: uri.query <> "&" <> query}
     end
   end
+
+  @doc """
+  Appends `path` to the given `uri`.
+
+  Path must start with "/" and cannot contain additional url components like
+  fragments or query strings.
+
+  ## Examples
+
+      iex> URI.append_path(URI.parse("http://example.com/foo/?x=1"), "/my-path") |> URI.to_string()
+      "http://example.com/foo/my-path?x=1"
+
+      iex> URI.append_path(URI.parse("http://example.com"), "my-path")
+      ** (ArgumentError) path must start with "/", got: "my-path"
+  """
+  @doc since: "1.15.0"
+  @spec append_path(t(), binary()) :: t()
+  def append_path(%URI{}, "//" <> _ = path) do
+    raise ArgumentError, ~s|path cannot start with "//", got: #{inspect(path)}|
+  end
+
+  def append_path(%URI{} = uri, "/" <> _ = path) do
+    current_path = String.trim_trailing(uri.path || "", "/")
+
+    case new(path) do
+      {:ok, %URI{fragment: nil, path: parsed_path, query: nil}}
+      when is_binary(parsed_path) ->
+        %{uri | path: current_path <> path}
+
+      {:ok, uri} ->
+        non_path_segment_message =
+          uri
+          |> Map.take([:query, :fragment])
+          |> Enum.reject(&match?({_, nil}, &1))
+          |> Enum.map(fn {k, v} -> "    #{k}: #{inspect(v)}" end)
+          |> Enum.join("\n")
```

Comment:
Nitpick: one less call

```suggestion
          |> Enum.map_join("\n", fn {k, v} -> "    #{k}: #{inspect(v)}" end)
```